#1949: Add documentation about commandlets and UrlUpdaters (#1974)
devonfw/IDEasy: documentation/contributing/commandlet-and-url-updater.adoc

* `WebsiteUrlUpdater`: Last resort, web scraping only (avoid if possible!)

=== JsonUrlUpdater (Recommended for REST APIs)

Use `JsonUrlUpdater` when you have access to a REST API that returns JSON with version information.

**Examples of tools using JsonUrlUpdater:**

* JavaUrlUpdater
* AndroidStudioUrlUpdater
* IdeaBasedUrlUpdater: which is used for IntelliJ, Pycharm

**What you need to implement:**

* `getTool()` - Return the tool name
* `doGetVersionUrl()` - Return the API endpoint URL
* `getJsonObjectType()` - Return the root JSON class (create with Jackson annotations)
* `getVersionItems(T)` - Extract version items from the JSON object
* `addVersion(UrlVersion, V)` - Process each version and call `doAddVersion()` with the download URL

=== GitHub Release Updater (Preferred for GitHub-hosted Tools)

Use `GithubUrlReleaseUpdater` when:

* The tool is hosted on GitHub
* Releases are published using GitHub's Release feature

**Examples of tools using GithubUrlReleaseUpdater:**

* Gradle
* Maven
* Mvnd
* Claude
* SquirrelSQL
* Inso

**How to identify if a tool uses GitHub Releases:**

1. Check if the GitHub repository has a "Releases" section

> 3. Example: https://api.github.com/repos/anthropics/claude-code/releases

Reviewer: Numbers do not make sense.

```suggestion
* Check if the GitHub repository has a "Releases" section
* Example: https://api.github.com/repos/anthropics/claude-code/releases
```

Use tags in your commandlet constructor:

[source,java]
----
public MyCommandlet(IdeContext context) {
  super(context, "mycommandlet", Set.of(Tag.JAVA, Tag.BUILD)); // Use one or more tags
}
----

== Package Manager Ecosystems

For tools that are part of established package manager ecosystems, IDEasy provides specialized base classes.
These ecosystems handle version discovery automatically via the package manager's registry and **no UrlUpdater is needed**.

=== Maven Ecosystem

**When to use:** For tools distributed via Maven Central Repository (or other Maven repositories)

To check if an artifact is available on Maven Central, visit https://mvnrepository.com/

**Base class:** `MvnBasedLocalToolCommandlet`

=== npm Ecosystem

**When to use:** For JavaScript/Node.js tools distributed via npm registry (npmjs.org)

To check if a package is available on npm, visit https://www.npmjs.com/

**Base classes:**

* `NodeBasedCommandlet`: For Node.js itself
* `NpmBasedCommandlet`: For npm packages

> 

Reviewer 2: Maybe this needs some clarification; I can see that there is a `NodeBasedCommandlet `(which only `NpmBasedCommandlet `seems to inherit directly?) and also there is a `Node.java` `LocalToolCommandlet `also seemingly to allow the usage of Node.js. What exactly is the difference here? 😅

Reviewer 2: Maybe we could include some examples in here that help to understand the difference between the three😊

Author: When implementing a new npm-based tool, you extend `NpmBasedCommandlet` (not `NodeBasedCommandlet` directly). I added a short sentence about it.

Reviewer 2: now it makes sense, thanks

The file format and name depend on the base class used.

To test the UrlUpdater, you can run in addition to the unit test the `UpdateInitiator` class located in `url-updater/src/main/java/com/devonfw/tools/ide/url/updater/UpdateInitiator.java` which allows you to run the updater and see the discovered versions and URLs in the console.
The arguments to pass are:

* Path to ide-urls project (where the url-updater module is located)
* Time to wait for the updater to finish (e.g., PT1H)
* Name of the updater to run (e.g., mytool)

Once a day, the `UpdateInitiator` will run the updaters through GitHub Actions and fill the https://github.com/devonfw/ide-urls[ide-URL] repository with new version/tools that are identified by the Updaters.

=== Create a Commandlet for the new tool

After merging the changes for the created UrlUpdater (if needed), you can check in https://github.com/devonfw/ide-urls if the new tool versions have been discovered and added to the `ide-urls` repository.

Then you can proceed to create the Commandlet for the new tool:

> * Commandlet class: `MyTool.java` located in: `cli/src/main/java/com/devonfw/tools/ide/cli/command/tool/mytool/`

Reviewer 2: ```suggestion
* Commandlet class: `MyTool.java` located in: `cli/src/main/java/com/devonfw/tools/ide/cli/tool/mytool/`
```

Author: yes that's true

* Extend the appropriate base class (`ToolCommandlet`, `LocalToolCommandlet`, or `GlobalToolCommandlet`)
* Use the tags defined in `Tag.java` to classify your tool
* Add the new tool to `cli/src/main/java/com/devonfw/tools/ide/commandlet/CommandletManagerImpl.java` by adding it to the `commandlets` list in the constructor.

> * Add a test for the created Commandlet in `src/test/java/com/devonfw/tools/ide/cli/command/tool/mytool/` with name `MyToolTest.java`

Reviewer 2: ```suggestion
* Add a test for the created Commandlet in `src/test/java/com/devonfw/tools/ide/cli/tool/mytool/` with name `MyToolTest.java`
```

Author: yes that's true

* For the test some resource files are needed in `src/test/resources/com/devonfw/tools/ide/cli/command/tool/mytool/` (depends on the base class used.

> For an example of the required files, you can check the tests of other commandlets with the same base class).

Reviewer 2: I cannot find this folder for other commandlets, is this new or did you maybe mean another folder?

<img width="361" height="327" alt="Image" src="https://github.com/user-attachments/assets/ab7e4871-34e2-4fb1-a1dd-9903d63deece" />

Author: yes i mean cli/src/test/resources/ide-projects/mytool

> * Proper help texts for all supported languages are added in `cli/src/main/resources/help/` (`help_en.properties`, `help_de.properties`).

Reviewer 2: On my local instance, they are under /main/resources/nls/* and they are named Help_de.properties and Help.properties (for english)

```suggestion
* Proper help texts for all supported languages are added in `cli/src/main/resources/nls/` (`Help.properties`, `Help_de.properties`).
```

Author: yes that's true

= Creating New Commandlets and UrlUpdaters

== Overview

> This document explains how to implement new commandlets and UrlUpdaters for IDEasy.

Reviewer: ```suggestion
This document explains how to implement a new `Commandlet` and `UrlUpdater` for IDEasy.
```

> A commandlet is a CLI command that manages a tool (install, uninstall, run), while a UrlUpdater is responsible for discovering and maintaining available versions of a tool.

Reviewer: ```suggestion
A `Commandlet` is a CLI sub-command and in case of a `ToolCommandlet` manages a tool (install, uninstall, run), while an `UrlUpdater` is responsible for discovering and maintaining available versions of a tool to sync the download-metadata in [ide-urls](http://github.com/ide-urls).
```

*** `GithubUrlReleaseUpdater`: Preferred for GitHub releases
*** `GithubUrlTagUpdater`: Fallback if no releases available
* `MavenBasedUrlUpdater`: For Maven repository tools
* `NpmBasedUrlUpdater`: For npm registry tools
* `WebsiteUrlUpdater`: Last resort, web scraping only (avoid if possible!)

=== JsonUrlUpdater (Recommended for REST APIs)

Use `JsonUrlUpdater` when you have access to a REST API that returns JSON with version information.

**Examples of tools using JsonUrlUpdater:**

* JavaUrlUpdater
* AndroidStudioUrlUpdater
* IdeaBasedUrlUpdater: which is used for IntelliJ, Pycharm

**What you need to implement:**

* `getTool()` - Return the tool name
* `doGetVersionUrl()` - Return the API endpoint URL
* `getJsonObjectType()` - Return the root JSON class (create with Jackson annotations)
* `getVersionItems(T)` - Extract version items from the JSON object
* `addVersion(UrlVersion, V)` - Process each version and call `doAddVersion()` with the download URL

=== GitHub Release Updater (Preferred for GitHub-hosted Tools)

Use `GithubUrlReleaseUpdater` when:

* The tool is hosted on GitHub
* Releases are published using GitHub's Release feature

> * Release assets contain downloadable binaries

Reviewer: can then also be removed

```suggestion
```

**Examples of tools using GithubUrlReleaseUpdater:**

* Gradle
* Maven
* Mvnd
* Claude
* SquirrelSQL
* Inso

**How to identify if a tool uses GitHub Releases:**

1. Check if the GitHub repository has a "Releases" section

> 2. Verify that releases contain downloadable artifacts (e.g., .zip, .tar.gz files)

Reviewer: This is not a requirement. We can derive the download URL also as a pattern and simply fill in the version as variable. There are some UrlUpdaters doing that using GithubUrlReleaseUpdater and there are no release downloads available on the github releases page of that tool.

```suggestion
```

3. Example: https://api.github.com/repos/anthropics/claude-code/releases

=== GitHub Tag Updater (When No Releases Available)

Use `GithubUrlTagUpdater` only as a fallback when:

* The tool doesn't use GitHub Releases

> * Releases are not available as downloadable artifacts

Reviewer: Can also be removed
```suggestion
```
